src/evaluation/retrieval.py

```diff
@@ -1,80 +1,106 @@
 import open_clip
 import numpy as np
 import torch
+from einops import rearrange
 
-
-def retrieval_evaluation(model_video, model_text, data, multicaption=False):
+def retrieval_evaluation(model_video, model_text, data):
     if type(data) == dict:
         dataloader = data["val"].dataloader
     else:
         dataloader = data
     device = "cuda" if torch.cuda.is_available() else "cpu"
     all_video_features, all_text_features = [], []
-    ground_truth = []
-    samp = 0
-
+    max_txt_len = 1
+    
     with torch.no_grad():
         for i, batch in enumerate(dataloader):
             embeddings = batch["embeddings"]
             toks = []
             # TODO: does this require batch_size = 1 ??
             for cap in batch["text"]:
-                if multicaption:
-                    for c in cap.split(";"): # multiple captions separated by ;
-                        toks.append(open_clip.tokenize(c))
-                        ground_truth.append(samp)
-                else:
-                    toks.append(open_clip.tokenize(cap))
-                    ground_truth.append(samp)
-                samp += 1
+              caps_list = cap.split(';')
+              if len(caps_list) > max_txt_len:
+                max_txt_len = len(caps_list)
+
+              for c in caps_list: # multiple captions separated by ;
+                  toks.append(open_clip.tokenize(c))
+          
             toks = torch.cat(toks)
             embeddings = embeddings.to(device, non_blocking=True)
             toks = toks.to(device, non_blocking=True)
 
             video_embeddings = model_video(embeddings)
-            text_embeddings = model_text(toks).float()
+            text_embeddings = model_text(toks)
 
             all_video_features.append(video_embeddings.cpu())
             all_text_features.append(text_embeddings.cpu())
-
+        dim_model = all_video_features[0].shape[-1]
+        text_features = zero_pad_text_features(all_text_features, max_txt_len, dim_model=dim_model)
+        video_features = torch.cat(all_video_features)
+        
         val_metrics = get_metrics(
-            video_features=torch.cat(all_video_features),
-            text_features=torch.cat(all_text_features),
-            ground_truth=ground_truth,
+            video_features=video_features,
+            text_features=text_features,
             logit_scale=100.0,
         )
     return val_metrics
 
+def zero_pad_text_features(text_features, max_txt_len, dim_model=512):
+  out = []
+  for mat in text_features:
+    padded = torch.zeros(size=(max_txt_len, dim_model))
+    padded[0:len(mat), :] = mat
+    
+    out.append(padded)
 
-def get_metrics(video_features, text_features, ground_truth, logit_scale):
-    metrics = {}
+  return torch.stack(out)
```

Comment:
you want to be explicit about what your functions do. This is "zero_pad_text_features" so the output should be the input but zero-padded, there's no mention of stacking so I'd just put that part back into the function that calls this.

Reply:
the thing is that the input is a torch tensor but `out` is a list. torch stacking `out` make it so `out` is the same type as the `text_features`, i.e. it actually is just a zero padded tensor.

Reply:
https://github.com/danielmend/temporal-embedding-aggregation/blob/2a3bc2066f383acc386450dd7a7c8db17ec7dd9b/src/evaluation/retrieval.py#L38
input is all_text_features which is a list since we append to it

Reply:
i dont even remember how my own code works 😂  ok on it.

Reply:
all good, I'd just wrap the zero_pad_text_features in torch.stack instead of wrapping out in torch.stack